### PR TITLE
[Model] Add experimental MammothModa2 distributed layerwise offload

### PR DESCRIPTION
**Pull request by 0z5a**
## Purpose

Add default-off, experimental distributed layerwise offload for MammothModa2-Preview through the existing `DistributedLayerwiseOffloadBackend`. An intermediate-layer failure can evict the first layer while leaving its slot ownership unchanged; the next request now re-materializes those weights before executing the group-first layer. The model's SP boundary also restores its padding/depth state when execution unwinds.

The updated branch integrates main `85da94339130337b49e610d08276aedcf84f6b96`
at head `9db73c2749499cc4b6ce8562f425083a28541fbb`, preserving the original
PR history. It still carries the shared MammothModa2 runtime/attention/SP
dependency work from #7160 (including #7134 and #7094). This remains an
independent DLO item under #7075, coordinated with #7158; it does not claim
completion of that issue's component/VAE offload work or change #7110's
norm/FFN ownership. No new fused norm/FFN kernel or checkpoint-key remapping
is included.

The DLO path:

- Registers one native ring over `gen_transformer.layers`, leaving refiners and VAE resident.
- Requires `extras.mammoth_experimental_dlo=true` and validates Preview, eager, one-request, Ulysses-only operation with zero leading resident blocks.
- Exercises both rank-local and AllGather transfers, including exact state restoration and subsequent requests after symmetric intermediate-layer failures.
- Adds full-checkpoint `Omni` AR → DiT → VAE A/B/A tests to the existing two-H100 nightly function-test route.
- Projects stage-local `cpu_offload_gb` and `cpu_offload_params` through the structured LLM load config. This uses vLLM's existing CPU weight offloader so AR and DiT can share smaller GPUs; it adds no model-specific CPU offload backend.

## Test Plan

Prerequisites: Linux, repository dependencies, two available CUDA GPUs with NCCL, and the official Preview checkpoint. Recorded runtime: Python 3.12.14, vLLM 0.28.0, Torch 2.13.0+cu130, Transformers 5.14.0 and Diffusers 0.40.0.

Native lifecycle/failure-recovery suite; no full checkpoint is needed:

```bash
HF_HUB_OFFLINE=0 TRANSFORMERS_OFFLINE=1 python -m pytest -q \
  --run-level=core_model -m cuda \
  tests/diffusion/distributed/test_mammothmoda2_dlo.py
```

Full-checkpoint local run. Put the pinned model at `$MODEL_PREFIX/bytedance-research/MammothModa2-Preview`; `MODEL_PREFIX` must include a trailing slash. On the tested 20 GiB GPUs, use the 12 GiB per-worker AR CPU offload budget shown here. Larger GPUs can leave `MAMMOTH_AR_CPU_OFFLOAD_GB` unset (default zero).

```bash
MODEL_PREFIX=/absolute/path/to/models/ \
MAMMOTH_AR_CPU_OFFLOAD_GB=12 CUDA_VISIBLE_DEVICES=0,1 \
HF_HUB_OFFLINE=1 TRANSFORMERS_OFFLINE=1 python -m pytest -sv \
  --run-level=full_model -m 'full_model and diffusion and cuda and cards_2' \
  tests/e2e/offline_inference/test_mammoth_moda2_expansion.py
```

CI route in `.buildkite/cuda/test-nightly.yml`:

```bash
pytest -sv \
  tests/e2e/online_serving/test_qwen_image_expansion.py \
  tests/e2e/offline_inference/test_mammoth_moda2_expansion.py \
  -m 'full_model and diffusion and H100 and B200 and cards_2' --run-level=full_model
```

## Test Result

The merge-only refresh at `9db73c2749499cc4b6ce8562f425083a28541fbb`
passed full diff-scoped pre-commit, compilation of all changed Python files,
and CPU adapter contracts. GPU and full-checkpoint results were not rerun.

**Fresh full-checkpoint qualification at `540446472ab3e105db89ea1bd2c18e14f908b29c`: 2 passed, no failures/errors/skips.** Both DLO modes completed all three A/B/A requests. Repeat A is pixel-exact in each mode, and all three corresponding outputs match exactly across rank-local and AllGather DLO. All 3,557 source files matched before and after the run. [Protocol and verified results](https://gist.github.com/0z5a/00c12d42d06f80ebf6357c58a2e59284).

The complete official checkpoint is pinned to `ef5a5e41dbf0de1ef6275586b7580f0d4248b4c6`; all 16 required files (37,067,585,024 bytes) were size/hash verified. The full-model run uses AR TP2, DiT Ulysses2/advanced UAA, BF16, 256×256 images, two DiT steps, seed 42 and guidance 1. Both workers' records verify the native 16-layer DLO ring and transport sizes. The full API tests cover repeated healthy requests; failure recovery is tested separately in the native suite because `Omni.generate` shuts down its engine after an inference exception.

The nine native two-rank cases passed on tree `82d40152881f455ccecdef020da53886e9ec551f`, with no skips: four lifecycle combinations, four symmetric intermediate-failure/recovery combinations, and one released-head-geometry fixture. Their 18 rank records contain 108 normal and 32 post-exception comparisons, all exact (`rtol=0, atol=0`), with exact restored state dicts. The production DLO and model files are unchanged between that tree and this follow-up. These fixtures use matched math SDPA; the full-checkpoint run uses the native FlashAttention path.

The structured configuration suite passed 173 tests with one skip after adding the LLM CPU offload fields. Applicable hooks and `git diff --check` passed. Earlier full-model attempts are retained: AR construction OOM before enabling CPU offload, rejected unowned CPU offload fields, and a test-only RPC reply-shape assertion after complete engine initialization. They are not counted as passes.

This is correctness and functional coverage. No new latency, throughput or image-quality result is claimed. Asymmetric collective/device failures, Dev T2I, concurrency/streaming, graphs, caching, quantization, LoRA, Host Weight Runtime/mmap, and other/mixed parallel layouts remain outside qualification. Historical performance measurements used older snapshots and are not results for this commit.

AI-assisted implementation and self-review; maintainer review and CI remain pending.

### CODE REVIEW

**Review by BLANKETusers — COMMENTED** [review]
The AR-to-diffusion adapter now rebuilds the downstream request with only `height`, `width`, and `additional_information`. However, the new DiT path reads seed, guidance, steps, and `cfg_range` from the diffusion request's `sampling_params`, and no longer reads the old values from `additional_information`. As a result, a normal T2I request with non-default sampling controls can reach the DiT with defaults instead of the user-supplied values. Could we forward the supported sampling fields here (including `cfg_range`), and add an integration test covering `ar2diffusion -> DiffusionRequestBatch -> _parse_request`?

File: vllm_omni/model_executor/stage_input_processors/mammoth_moda2.py [hunk 1]
```diff
@@ -25,91 +19,65 @@ def _as_dict(prompt: Any) -> dict[str, Any]:
...
-                additional_information=additional_information,
-                multi_modal_data=None,
-                mm_processor_kwargs=None,
-            )
-        )
-
-    return dit_inputs
+            "answer_start_index": len(prompt_token_ids),
+        },
+    }
```
Comment (BLANKETusers) [inline vllm_omni/model_executor/stage_input_processors/mammoth_moda2.py:83]: The AR-to-diffusion adapter now rebuilds the downstream request with only `height`, `width`, and `additional_information`. However, the new DiT path reads seed, guidance, steps, and `cfg_range` from the diffusion request's `sampling_params`, and no longer reads the old values from `additional_information`. As a result, a normal T2I request with non-default sampling controls can reach the DiT with defaults instead of the user-supplied values. Could we forward the supported sampling fields here (including `cfg_range`), and add an integration test covering `ar2diffusion -> DiffusionRequestBatch -> _parse_request`?

Comment (0z5a) [reply vllm_omni/model_executor/stage_input_processors/mammoth_moda2.py:83]: Added the requested bridge coverage in [0d5f0d64](https://github.com/0z5a/vllm-omni/commit/0d5f0d64a4401d85cec5320d8971256c036a51cf). It uses the real adapter, diffusion-client serializer, request normalization, batch and parser, checking non-default controls, model-specific overrides, explicit zero values, and exact conditioning/token alignment.

Tracing the current path, the adapter builds conditioning while the orchestrator/stage pool pass the next stage's sampling parameters independently. The client serializes `seed`, `guidance_scale`, `num_inference_steps` and `extra_args` (including `cfg_range`); serving normalizes the declared model-specific controls into `extra_args`. I therefore have not reproduced a standard-field drop or changed production forwarding.

All 3 new CPU cases pass; the neighboring model/adapter suite is 118 passed, 4 deselected (including those 3), and applicable hooks pass. This is CPU bridge coverage, not live orchestrator/RPC, AR-model or GPU inference. If the failing caller stores controls only in legacy `additional_information`, could you share that entry point/payload so we can qualify that compatibility path separately?

File: vllm_omni/diffusion/models/mammoth_moda2/pipeline_mammothmoda2_dit.py [hunk 2]
```diff
@@ -35,8 +156,13 @@ class MammothModa2DiTPipeline(nn.Module, SupportsComponentDiscovery):
     _dit_modules: ClassVar[list[str]] = ["gen_transformer"]
     _encoder_modules: ClassVar[list[str]] = ["gen_image_condition_refiner"]
     _vae_modules: ClassVar[list[str]] = ["gen_vae"]
+    # One circular ring, executed for every conditional/unconditional pass.
+    # Auxiliary refiners and VAE remain resident; no encoder/mmap capability
+    # or leading-resident-layer lifecycle is claimed by this experiment.
+    _offload_plan: ClassVar[OffloadPlan] = OffloadPlan(block_attrs={"gen_transformer": ("layers",)})
```
Comment (BLANKETusers) [inline vllm_omni/diffusion/models/mammoth_moda2/pipeline_mammothmoda2_dit.py:162]: This enables distributed layerwise offload for MammothModa2, but the inference error path does not restore the ring state before the worker accepts the next request. If a later layer fails after the first layer has been offloaded but before the ring prefetches the first layer again, the first layer can remain backed by placeholder storage. The next request may enter the group-first path and skip re-materialization because the slot still belongs to the same group. Could we restore or re-materialize the offload ring when execution fails, so a failed request does not poison subsequent requests?

Comment (0z5a) [reply vllm_omni/diffusion/models/mammoth_moda2/pipeline_mammothmoda2_dit.py:162]: Fixed in [54044647](https://github.com/0z5a/vllm-omni/commit/540446472ab3e105db89ea1bd2c18e14f908b29c). The group-first hook now synchronously reloads when `is_materialized` is false, even when the shared slot still belongs to that group. An intermediate-layer exception therefore cannot make the next request treat placeholder storage as resident weights. The model's SP boundary also unwinds its padding/depth state on exceptions.

Validation on two RTX 4000 Ada Generation GPUs (20 GiB each):

- **Nine native two-rank cases passed**, including FP32/BF16 × rank-local/AllGather symmetric intermediate-layer failures in a three-layer ring. Two subsequent requests reuse the same enabled backend and hook group, for guidance 1 and 4. Both ranks match an independent resident baseline exactly (`rtol=0, atol=0`); disable restores the full state dict exactly. These ran on tree `82d40152`; the production DLO and model files are unchanged in this follow-up.
- **Two full-checkpoint E2E cases passed, with no failures or skips.** Full official Preview checkpoint `ef5a5e41dbf0de1ef6275586b7580f0d4248b4c6`, native `Omni` AR → DiT → VAE, separately for rank-local and AllGather DLO. Each mode runs A/B/A on the same engine, checks both workers' actual 16-layer DLO ring/transport, and checks nonconstant RGB images, different A/B output and repeat-A agreement. Repeat A has a maximum RGB difference of zero in both modes; the three corresponding rank-local/AllGather images also match exactly.
- The stage-local CPU offload configuration suite passed **173 tests, one skipped**, and applicable hooks passed. All 3,557 source files matched this commit’s tree before and after the full E2E run. [Verified protocol and result records](https://gist.github.com/0z5a/00c12d42d06f80ebf6357c58a2e59284).

The full-model protocol uses AR TP2, DiT Ulysses2/advanced UAA, BF16, 256×256, two DiT steps and guidance 1. AR uses vLLM's existing CPU weight offloader (`12 GiB`, `mlp` selection) to share these smaller GPUs; the PR now projects those arguments through the structured LLM load config. The new E2E cases are included in the existing two-H100 nightly function-test route. The earlier OOM, unowned-config and RPC-result-shape attempts are retained in the validation record.

This validates full-model repeated requests and, separately, symmetric failure recovery with a healthy NCCL group. `Omni.generate` shuts down its engine on an inference exception, so the full API test does not claim same-engine recovery after an injected failure. Asymmetric collective/device failures and performance or image-quality qualification remain outside this result.

Prepared with Codex assistance.

**Comment by hsliuustc0106** [comment]
let's wait for the basic kernel optimization PR done
